### PR TITLE
Fix: Enhance parser "SatellitePostgreSQLQuery"

### PR DESCRIPTION
**Pull request by huali027**
Signed-off-by: Huanhuan Li <huali@redhat.com>

### All Pull Requests:

Check all that apply:

* [ ] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [ ] Is this PR to correct an issue?
* [x] Is this PR an enhancement?

### CODE REVIEW

**Comment by huali027** [comment]
Fixes: https://github.com/RedHatInsights/insights-core/issues/3313

**Comment by huali027** [comment]
@xiangce Please help to review, thanks.

**Comment by clarabez** [comment]
Hi, @huali027!
We've identified that setup pipeline is failing for RHEL6 on "QE Pull Request Testing", but it is working as expected for RHEL79, RHEL84 and RHEL90.

This issue has been reported and is being analyzed here: [RHCLOUD-17392](https://issues.redhat.com/browse/RHCLOUD-17392).

File: insights/parsers/satellite_postgresql_query.py [hunk 1]
```diff
@@ -72,15 +72,24 @@ class SatellitePostgreSQLQuery(CommandParser, list):
...
         SkipException: when there isn't data in the table
-        ParseException: when the output isn't in good csv format
+        ParseException: when the output isn't in good csv format or the subclass doesn't define headers
     """
+    headers = []
 
     def parse_content(self, content):
-        if not content or len(content) == 1:
-            raise SkipException("There is no data in the table")
+        if self.__class__.__name__ != 'SatellitePostgreSQLQuery' and not self.headers:
```
Comment (xiangce) [inline insights/parsers/satellite_postgresql_query.py:80]: When defining the base class like this (use an undefined/empty static member as a required `gurad`), it's not a good idea to instantiate or use it like this. that's to say, this base class should be instantiated due the `self.headers` is never contain a value, unless in its child classes. 

Better remove all the direct tests against the base class. but test them in the child-classes instead.

File: insights/parsers/satellite_postgresql_query.py [hunk 2]
```diff
@@ -139,6 +148,7 @@ class SatelliteAdminSettings(SatellitePostgreSQLQuery):
         >>> table.get_setting('destroy_vm_on_host_delete')
         True
     """
+    headers = ['name', 'value', 'default']
```
Comment (xiangce) [inline insights/parsers/satellite_postgresql_query.py:151]: This is different with the description of Line-163

File: insights/parsers/satellite_postgresql_query.py [hunk 3]
```diff
@@ -72,15 +72,24 @@ class SatellitePostgreSQLQuery(CommandParser, list):
...
+        ParseException: when the output isn't in good csv format or the subclass doesn't define headers
     """
+    headers = []
 
     def parse_content(self, content):
-        if not content or len(content) == 1:
-            raise SkipException("There is no data in the table")
+        if self.__class__.__name__ != 'SatellitePostgreSQLQuery' and not self.headers:
+            raise ParseException("Please define the headers attribute which contains the column names.")
+        start_index = None
```
Comment (xiangce) [inline insights/parsers/satellite_postgresql_query.py:82]: the `insights.parsers.calc_offset` can be used to get the `start_index`.

**Review by xiangce — CHANGES_REQUESTED** [review]

File: insights/parsers/satellite_postgresql_query.py (hunk 2 above)
Comment (huali027) [reply insights/parsers/satellite_postgresql_query.py:122]: Do u mean I don't mention "name" in Line-163? In Line-163, I just explain "default" and "value" must both be queried to determine the final value of the setting.

File: insights/parsers/satellite_postgresql_query.py (hunk 1 above)
Comment (huali027) [reply insights/parsers/satellite_postgresql_query.py]: @xiangce I update the code to move the test in child-classes, please help to review again, thanks.

File: insights/parsers/satellite_postgresql_query.py (hunk 3 above)
Comment (huali027) [reply insights/parsers/satellite_postgresql_query.py]: Updated, please help to check again, thanks.

File: insights/parsers/satellite_postgresql_query.py (hunk 2 above)
Comment (xiangce) [reply insights/parsers/satellite_postgresql_query.py:122]: By defining the `header`, I think it implicitly means that the header line of the content should contain all of the items in the `header`.  so the L145~L146 is not required anymore.

And the base class should also be updated as:

```
if not all(t in content[start_index] for t in self.headers):
   raise ...
```

Comment (xiangce) [reply insights/parsers/satellite_postgresql_query.py:122]: I just enhanced the `calc_offset` and added a new `require_all` parameter
- https://github.com/RedHatInsights/insights-core/pull/3316
After that, you can use it, like:
```
start_index = calc_offset(content, self.headers, require_all=True)
```

**Comment by huali027** [comment]
@xiangce Please help to review, thanks.

**Review by xiangce — APPROVED** [review]

**Pull request merged**